pkg/controller/ttlafterfinished/ttlafterfinished_controller.go

```diff
@@ -20,92 +20,107 @@ import (
+	// Resources that the controller will check its TTL and attempt to delete when the TTL expires.
 	queue workqueue.RateLimitingInterface
 
 	// The clock for tracking time
 	clock clock.Clock
 }
 
-// New creates an instance of Controller
-func New(jobInformer batchinformers.JobInformer, client clientset.Interface) *Controller {
+// New creates an instance o Controller
+func New(informerFactory informers.SharedInformerFactory, resource schema.GroupVersionResource, dynamicClient dynamic.Interface, client clientset.Interface) *Controller {
 	eventBroadcaster := record.NewBroadcaster()
 	eventBroadcaster.StartLogging(klog.Infof)
 	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: client.CoreV1().Events("")})
 
 	if client != nil && client.CoreV1().RESTClient().GetRateLimiter() != nil {
-		metrics.RegisterMetricAndTrackRateLimiterUsage("ttl_after_finished_controller", client.CoreV1().RESTClient().GetRateLimiter())
+
+		metrics.RegisterMetricAndTrackRateLimiterUsage(fmt.Sprintf("ttl_%v_after_finished_controller", resource.Resource), client.CoreV1().RESTClient().GetRateLimiter())
 	}
 
 	tc := &Controller{
-		client:   client,
-		recorder: eventBroadcaster.NewRecorder(scheme.Scheme, v1.EventSource{Component: "ttl-after-finished-controller"}),
-		queue:    workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "ttl_jobs_to_delete"),
+		client:   dynamicClient.Resource(resource),
+		recorder: eventBroadcaster.NewRecorder(scheme.Scheme, v1.EventSource{Component: fmt.Sprintf("ttl-%v-after-finished-controller", resource.Resource)}),
+		resource: resource,
+		queue:    workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), fmt.Sprintf("ttl_%s_to_delete", resource.Resource)),
+	}
+
+	genericInformer, err := informerFactory.ForResource(resource)
+	if err != nil {
+		return nil
 	}
 
-	jobInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
-		AddFunc:    tc.addJob,
-		UpdateFunc: tc.updateJob,
+	tc.rInformer = genericInformer
+
+	tc.rInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    tc.addResource,
+		UpdateFunc: tc.updateResource,
 	})
 
-	tc.jLister = jobInformer.Lister()
-	tc.jListerSynced = jobInformer.Informer().HasSynced
+	tc.rListerSynced = tc.rInformer.Informer().HasSynced
 
 	tc.clock = clock.RealClock{}
 
 	return tc
 }
 
-// Run starts the workers to clean up Jobs.
+// Run starts the workers to clean up Jobs/Pods.
 func (tc *Controller) Run(workers int, stopCh <-chan struct{}) {
 	defer utilruntime.HandleCrash()
 	defer tc.queue.ShutDown()
 
-	klog.Infof("Starting TTL after finished controller")
+	klog.Infof("Starting TTL after finished controller for %s", tc.resource.String())
 	defer klog.Infof("Shutting down TTL after finished controller")
 
-	if !controller.WaitForCacheSync("TTL after finished", stopCh, tc.jListerSynced) {
+	if !controller.WaitForCacheSync("TTL after finished", stopCh, tc.rListerSynced) {
 		return
 	}
 
```

Comment:
We could just run a dedicated worker for Pods, and a dedicated queue/informer/lister for Pods too. 

We could use something more generic like what [garbage collector](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/garbagecollector/garbagecollector.go) uses, but that seems to be premature optimization. We can move to use that if we're supporting custom resources.

Reply:
@janetkuo Thanks. I’ll refactor the code with dedicated queue/informer/lister for pods and jobs

Out of curiosity, can I ask why not use two controllers for jobs and pods if we are using dedicated queues/informers/lister(which I think is the crux of a controller) anyways? Am I missing anything?